Data Structures/Array/amortisation.cpp

```diff
@@ -0,0 +1,99 @@
+// Name : Shivam Singhal
+// Email ID : shivamsinghal0610@gmail.com
+ 
+// Array doubling and amortised complexity
+ 
+#include <iostream>
+using namespace std;
+ 
+// function to double array size using amortised technique
+void array_doubling(int c, int arr[], int* size)
+{
+	//doubling array size (using call by reference)
+	*size = (*size) * 2;
+	// creating temporary array of double size
+	int double_arr[*size];
+	// copying existing elements in original array to temporary array
+	for(int i = 0; i < c; i++)
+		double_arr[i] = arr[i];
+	// deleting memory allocated by original array
+	delete [] arr;
+	// re-inserting values into original array with double size
+	for(int i = 0; i < c; i++)
+		arr[i] = double_arr[i];
+}
+ 
+int main(){
+	int c = 0;
+	int size = 1;
+	//allocating original array of size 1
+	int* arr = new int[size];
+	while(1){
+		cout << "Please enter a number.\n";
+		int n;
+		cin >> n;
+		cout << n << endl;
+		//if index at which new number is to be entered doesn't exceed limit
+		if (c < size)
+			arr[c++] = n;
+		else{
+			//doubling array using amortise technique
+			array_doubling(c, arr, &size);
+			//entering new number
+			arr[c++] = n;
+		}
+		cout << "Do you wish to enter another number ? (Y, y / N, n)\n";
+		char p;
+		cin >> p;
+		cout << p << endl;
+		if (p == 'N' || p == 'n')
+			break;
+	}
+	cout << "Final size of array : " << size << endl;
+	cout << "Number of elements stored in array : " << c << endl;
+	cout << "Array elements : " << endl;
+	for(int i = 0; i < c; i++)
+		cout << arr[i] << " ";
+	return 0;
+}
+// Sample Case :
+// Input : 
+// 1
+// Y
+// 2
+// N
+ 
+// Output :
+// Please enter a number.
+// 1
+// Do you wish to enter another number ? (Y, y / N, n)
+// Y
+// Please enter a number.
+// 2
+// Do you wish to enter another number ? (Y, y / N, n)
+// N
+// Final size of array : 2
+// Number of elements stored in array : 2
+// Array elements : 
+// 1 2 
+ 
+ 
+ 
+//Time Complexity for insertion : O(1)
```

Comment:
It will be good if u also write worst case time complexity.

Reply:
Done.